packages/block-library/src/navigation-link/update-attributes.js

```diff
@@ -89,7 +89,7 @@ export const updateAttributes = (
 
 	setAttributes( {
 		// Passed `url` may already be encoded. To prevent double encoding, decodeURI is executed to revert to the original string.
-		...( newUrl && { url: encodeURI( safeDecodeURI( newUrl ) ) } ),
+		...{ url: newUrl ? encodeURI( safeDecodeURI( newUrl ) ) : newUrl },
```

Comment:
I'm still not super clear why it was necessary to `encodeURI()` here in the first place. It seems like `https://example.com?s=<>` is JSON-encoded just fine:

```
>> var test = {};
undefined
>> test.url = 'https://example.com?s=<>';
"https://example.com/?s=<>"
>> JSON.stringify(test);
"{\"url\":\"https://example.com/?s=<>\"}"
```

Maybe because the `<>` gets stripped out by kses? Can you debug and document why https://github.com/WordPress/gutenberg/pull/19679 was necessary in the first place? If it's not necessary, maybe we can remove it entirely.

Reply:
> The reason that this was stripped out is because of `esc_url` function that uses `kses` under the hood.

@kozer For posterity, can you provide a full step-by-step documentation of what's going on?

Reply:
For sure! So, the process is the following:

- User enters a new URL via the editor.
- Before `encodeURI`  introduced the URL was stored as is.
- After saving, when someone navigates to the page with the inserted URL, the URL is passed in the `render_block_core_navigation_link` function.
- When rendered, the URL is passed through `esc_url`. `esc_url` uses the `preg_replace`  function that seems to discard `<` and `>` characters:
```
	$url = preg_replace( '|[^a-z0-9-~+_.?#=!&;,/:%@$\|*\'()\[\]\\x80-\\xff]|i', '', $url );
```
- Using `encodeURI` in a URL that has those characters, (eg: `encodeURI('http:example.com?s=<>')` encodes it, and produce an encoded URL (in the above example the URL is encoded to `http:example.com?s=%3C%3E`, and so, `esc_url` no longer strips out those characters.

This is the reason why `encodeURI` was introduced originally in #19679, and produce the bug we are now facing.